File: js/bench2d.js
var Vec2 = Box2D.Common.Math.b2Vec2
  , BodyDef = Box2D.Dynamics.b2BodyDef
  , Body = Box2D.Dynamics.b2Body
  , FixtureDef = Box2D.Dynamics.b2FixtureDef
  , Fixture = Box2D.Dynamics.b2Fixture
  , World = Box2D.Dynamics.b2World
  , MassData = Box2D.Collision.Shapes.b2MassData
  , PolygonShape = Box2D.Collision.Shapes.b2PolygonShape
  , CircleShape = Box2D.Collision.Shapes.b2CircleShape
  ;

var world;

var PYRAMID_SIZE = 40;

function init() {
  var gravity = new Vec2(0, -10);
  world = new World(gravity, true);

  {
    var shape = new PolygonShape();
    shape.SetAsEdge(new Vec2(-40.0, 0), new Vec2(40.0, 0));

    var fd = new FixtureDef();
    fd.density = 0.0;
    fd.shape = shape;

    var bd = new BodyDef();
    var ground = world.CreateBody(bd);
    ground.CreateFixture(fd);
  }

  {
    var a = .5;
    var shape = new PolygonShape();
    shape.SetAsBox(a, a);

    var x = new Vec2(-7.0, 0.75);
    var y = new Vec2();
    var deltaX = new Vec2(0.5625, 1);
    var deltaY = new Vec2(1.125, 0.0);

    for (var i = 0; i < PYRAMID_SIZE; ++i) {
      y.Set(x.x, x.y);

      for (var j = i; j < PYRAMID_SIZE; ++j) {
        var fd = new FixtureDef();
        fd.density = 5.0;
        fd.shape = shape;

        var bd = new BodyDef();
        bd.type = Body.b2_dynamicBody;
        bd.position.Set(y.x, y.y);
        var body = world.CreateBody(bd);
        body.CreateFixture(fd);
        y.Add(deltaY);
      }

      x.Add(deltaX);
    }
  }
}

function step() {
  world.Step(1 / 60, 3, 3);
}

File: c/Bench2d.cpp
#include <cstdio>
#include <time.h>
#include <math.h>
#include <stdlib.h>

#include "Box2D/Box2D.h"
#include "Bench2d.h"

using namespace std;

const int e_count = 40;

// Simple nearest-rank %ile (on sorted array). We should have enough samples to make this reasonable.
float percentile(clock_t times[FRAMES], float pc) {
  int rank = (int)((pc * FRAMES) / 100);
  return times[rank];
}

int _cmp(const void *a, const void *b) {
	return *(clock_t*)a - *(clock_t*)b;
}

result_t measure(clock_t times[FRAMES]) {
  float values[FRAMES];
  result_t r;

	float total = 0;
	for (int i = 0; i < FRAMES; ++i) {
		values[i] = (float)times[i] / CLOCKS_PER_SEC * 1000;
	 	total += values[i];
	}
  r.mean = total / FRAMES;

	qsort(times, FRAMES, sizeof(clock_t), _cmp);
  r.pc_5th = percentile(times, 5) / CLOCKS_PER_SEC * 1000;
  r.pc_95th = percentile(times, 95) / CLOCKS_PER_SEC * 1000;
  return r;
}

result_t bench() {
	// Define the gravity vector.
	b2Vec2 gravity(0.0f, -10.0f);

	// Construct a world object, which will hold and simulate the rigid bodies.
	b2World world(gravity);
  world.SetAllowSleeping(false);

	{
		b2BodyDef bd;
		b2Body* ground = world.CreateBody(&bd);

		b2EdgeShape shape;
		shape.Set(b2Vec2(-40.0f, 0.0f), b2Vec2(40.0f, 0.0f));
		ground->CreateFixture(&shape, 0.0f);
	}

  b2Body* topBody;

	{
		float32 a = 0.5f;
		b2PolygonShape shape;
		shape.SetAsBox(a, a);

		b2Vec2 x(-7.0f, 0.75f);
		b2Vec2 y;
		b2Vec2 deltaX(0.5625f, 1);
		b2Vec2 deltaY(1.125f, 0.0f);

		for (int32 i = 0; i < e_count; ++i) {
			y = x;

			for (int32 j = i; j < e_count; ++j) {
				b2BodyDef bd;
				bd.type = b2_dynamicBody;
				bd.position = y;
				b2Body* body = world.CreateBody(&bd);
				body->CreateFixture(&shape, 5.0f);

        topBody = body;

				y += deltaY;
			}

			x += deltaX;
		}
	}

	for (int32 i = 0; i < WARMUP; ++i) {
		world.Step(1.0f/60.0f, 3, 3);
  }

	clock_t times[FRAMES]; 
	for (int32 i = 0; i < FRAMES; ++i) {
		clock_t start = clock();
		world.Step(1.0f/60.0f, 3, 3);
		clock_t end = clock();
		times[i] = end - start;
#if DEBUG
    printf("%f :: ", topBody->GetPosition().y);
		printf("%f\n", (float32)(end - start) / CLOCKS_PER_SEC * 1000);
#endif
	}

  return measure(times);
}
